pennylane/transforms/add_noise.py

```diff
@@ -0,0 +1,318 @@
+        noise1 = qml.noise.partial_wires(qml.PhaseDamping, 0.4)
+
+        fcond2 = qml.noise.op_in([qml.RX, qml.RZ])
+        def noise2(op, **kwargs):
+            qml.ThermalRelaxationError(op.parameters[0] * 0.5, kwargs["t1"],  kwargs["t2"], 0.6, op.wires)
+
+        noise_model = qml.NoiseModel({fcond1: noise1, fcond2: noise2}, t1=2.0, t2=0.2)
+
+        @partial(qml.transforms.add_noise, noise_model=noise_model)
+        @qml.qnode(dev)
+        def circuit(w, x, y, z):
+            qml.RX(w, wires=0)
+            qml.RY(x, wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.RY(y, wires=0)
+            qml.RX(z, wires=1)
+            return qml.expval(qml.Z(0) @ qml.Z(1))
+
+    Executions of this circuit will differ from the noise-free value:
+
+    >>> circuit(0.9, 0.4, 0.5, 0.6)
+    tensor(0.60722291, requires_grad=True)
+    >>> print(qml.draw(f)(0.9, 0.4, 0.5, 0.6))
+    0: ──RX(0.9)──PhaseDamping(0.4)───────────────────────╭●──RY(0.5)───ThermalRelaxationError(0.2,2.0,0.2,0.6)─┤ ╭<Z@Z>
+    1: ──RY(0.4)──ThermalRelaxationError(0.2,2.0,0.2,0.6)─╰X──RX(0.6)───PhaseDamping(0.4)───────────────────────┤ ╰<Z@Z>
+
+    .. details::
+        :title: Usage Details: Advanced Examples
+        :href: add-noise-example
+
+        ``qml.add_noise`` can also be used to tranform quantum functions, devices or already constructed qnodes.
+        To do this, one may construct noise models with various kinds of ``conditionals`` and ``noise_fns`` key-value pairs.
+        For example, for a list of specified operations, one can functionally add noise operations for their every encounter
+        in the circuit based on their wires:
+
+        .. code-block:: python3
+
+            fcond1 = qml.noise.op_in(["X", "Y"])
+            noise1 = qml.noise.partial_wires(qml.AmplitudeDamping, 0.4)
+
+        Moreover, a functional conditional construction could also be done in cases where more complex evaluations are needed
+        than the ones done by the already provided conditional constructors in the :func:`qml.noise <pennylane.noise>`.
+
+        .. code-block:: python3
+
+            @qml.BooleanFn
+            def fcond2(op, **kwargs):
+                return isinstance(op, qml.RY) and op.parameters[0] >= 0.5
+            noise2 = qml.noise.partial_wires(qml.PhaseDamping, 0.9)
+
+        By default, for each operation for which a conditional evaluates to ``True``, the corresponding noise operations gets
+        queued following an `iterative-insertion` approach. One could change this by including the operation itself, and
+        queue it with :func:`~.pennylane.apply` as given below:
+
+        .. code-block:: python3
+
+            fcond3 = qml.noise.op_eq(qml.RX) & qml.noise.wires_in([0, 1])
+            def noise3(op, **kwargs):
+                qml.RZ(op.parameters[0] * 0.05, op.wires)
+                qml.apply(op) # <-- sandwiched between two RZs
+                qml.RZ(-op.parameters[0] * 0.05, op.wires)
```

Comment:
About the option you provide - I did think about something like this, but in reality, this would still restrict users from doing something more flexible than our defined position.

Reply:
We could also use a tuple: `(noise: Callable, replace: Bool)`. If replace is True, then we pop initial operation and use the noise only. Then the user can do as much sandwiching as they want. In my opinion the interface is not clear for this case. 

However, this is not a blocker for this PR, we can revisit this advanced use case in the future. Just want to give some suggestions.

Reply:
@obliviateandsurrender does the current approach work? wouldn't it cause the op to be applied twice?

Reply:
Yep, the current approach works! For example, a sandwich example - 

<img width="971" alt="image" src="https://github.com/PennyLaneAI/pennylane/assets/23234279/805872f5-ecc1-4e18-86e0-7957d4056346">